pyro/contrib/forecast/forecaster.py

```diff
@@ -0,0 +1,220 @@
+        :type covariates: ~torch.Tensor
+        :returns: Return value is ignored.
+        """
+        raise NotImplementedError
+
+    @property
+    def time_plate(self):
+        """
+        :returns: A plate named "time" with size ``covariates.size(-2)`` and
+            ``dim=-1``. This is available only during model execution.
+        :rtype: :class:`~pyro.plate`
+        """
+        assert self._time_plate is not None, ".time_plate accessed outside of .model()"
+        return self._time_plate
+
+    def predict(self, noise_dist, prediction):
+        """
+        Prediction function, to be called by :meth:`model` implementations.
+        This should be called outside of the :meth:`time_plate`.
+
+        :param noise_dist: A noise distribution with ``.event_dim == 2``.
+            ``noise_dist`` is typically zero-mean or zero-median or zero-mode
+            or somehow centered.
+        :type noise_dist: ~pyro.distributions.Distribution
+        :param prediction: A prediction for the data. This should have the same
+            shape as ``data``, but extended to full duration of the
+            ``covariates``.
+        :type prediction: ~torch.Tensor
+        """
+        assert self._data is not None, ".predict() called outside .model()"
+        assert self._forecast is None, ".predict() called twice"
+        assert isinstance(noise_dist, dist.Distribution)
+        assert isinstance(prediction, torch.Tensor)
+        assert noise_dist.event_shape == prediction.shape[-2:]
+
+        # The following reshaping logic is required to reconcile batch and
+        # event shapes. This would be unnecessary if Pyro used name dimensions
+        # internally, e.g. using Funsor.
+        #
+        #     batch_shape                    | event_shape
+        #     -------------------------------+----------------
+        #  1. sample_shape + shape + (time,) | (obs_dim,)
+        #  2.           sample_shape + shape | (time, obs_dim)
+        #  3.    sample_shape + shape + (1,) | (time, obs_dim)
+        #
+        # Parameters like noise_dist.loc typically have shape as in 1.  However
+        # calling .to_event(1) will shift the shapes resulting in 2., where
+        # sample_shape+shape will be misaligned with other batch shapes in the
+        # trace. To fix this the following logic "unsqueezes" the distribution,
+        # resulting in correctly aligned shapes 3. Note the "time" dimension is
+        # effectively moved from a batch dimension to an event dimension.
+        noise_dist = reshape_batch(noise_dist, noise_dist.batch_shape + (1,))
+        data = self._data.unsqueeze(-3)
+        prediction = prediction.unsqueeze(-3)
+
+        # Create a sample site.
+        t_obs = data.size(-2)
+        t_cov = prediction.size(-2)
+        if t_obs == t_cov:  # training
+            pyro.sample("residual", noise_dist, obs=data - prediction)
+            self._forecast = data.new_zeros(data.shape[:-2] + (0,) + data.shape[-1:])
```

Comment:
why fill this during training?

Reply:
It seems safer to keep a consistent return type and contract that
```py
assert isinstance(result, torch.Tensor)
assert result.dtype == data.dtype
assert result.device == data.device
assert result.shape == data.shape[:-1] + covariates.shape[-2:-1] + data.shape[-1:]
```
Note this costs nothing; when training the tensor is empty `self._forecast.numel() == 0`.